brave/brave-ios
# Fix #3229: Implement Open Tabs Sync

## Summary of Changes 

This Pull Request contains changes

Tab Tray Open Tabs List:

- A segmented control in Tab Tray screen which allow user to choose between actual tab tray with tabs and a screen which shows synced sessions only in normal mode.
- A list of synced sessions from other devices in sync chain with expand/shrink functionality  and details of the synced session like name of the device and time period of last sync happened. 
- Clicking a synced session item will open that url in a new tab.
- A search functionality which will only present synced session information that includes the search keyword + an empty placeholder view when no search result found.

Sync Settings:

- Adding Open Tabs as a sync type + toggle on / off functionality for Open Tabs Sync type.
- A new placeholder view for Sync Settings for empty state when device list is empty, used instead of navigating back in controller.

Send / Receive Webpage:

- Add Send Webpage action to ... menu and as an activity among other share activities.
- Send Webpage which contains list of devices active in Sync Chain + details of the device like name and active period + selection to designated device + Send action which will send current url from the active page to another device.
- Observers checking if any remote entry is sent to device and displaying a toast with message and action button which will open the link in a new tab.



This pull request fixes #3229

Brave Core PR is merged and uplifted to 1.43.x and security review is completed for the core side.

Core PR For Open Tab Sync: https://github.com/brave/brave-core/pull/14287

Sec - Privacy Review: https://github.com/brave/security/issues/977

## Submitter Checklist:


## Test Plan - Screenshots: 




### General look at changes in screens:

Tab Tray Open Tabs List:

![test1](https://user-images.githubusercontent.com/6643505/183492513-33b8bb3b-3aea-43dd-b2ff-d7ed2a4ba0a1.png)

Sync Settings:

![test3](https://user-images.githubusercontent.com/6643505/183492703-b5349c9d-405d-4c93-bf47-7b2b4572ba61.png)

Send - Receive Webpage:

![test2](https://user-images.githubusercontent.com/6643505/183492730-5a824ea1-5840-4624-b3ad-a07c75f33460.png)


### Test Plan for Tab Tray Open Tabs List:

Initial Part before joining a sync chain and enabling Open Tabs toggle in settings.

- Open Tab Tray and check under the search bar there is a segmented control to choose Local Tab List and Open Tabs (Private Mode should not have this segmented control to view open tabs from different devices)
- By changing private versus normal mode check segmented control and open tabs screen can be viewed just only normal mode.
- Select Open Tab segment and check a view indicating empty state is shown
- Check Search Bar placeholder text is changed according to which segment user is actively monitoring
- Activate search bar and typing in search bar should populate empty search state at this stage

After testing initial part with no open tabs sync type enabled, second part will be joining a sync chain and enabling open tabs

- Navigate Settings -> General Section -> Sync (at this stage user either joins a new sync chain or view sync settings according to their sync status - this part is not changed in anyway except adding Open Tabs toggle as new sync type) (Also Open tabs will not be enabled automatically for a new user since the only default sync type is Bookmarks)
- At this point Enabling Open Tabs should start gathering Synced Sessions from other devices and sync normal mode tabs with the sync chain

After joining a sync chain and enabling Open Tabs

- Open Tab Tray and navigate to Open Tabs segment and check the synced session from other devices can be monitored.
- Every section on that list will be different synced session fetched from brave-core and clicking on section header should shrink/expand the list under section.
- At this point the name of the device and when it is last synced should be able monitored on every section including a icon indicating what kind or a device it is (either a mobile device or pc/laptop)
- Inside every synced session, a list of synced website entry information should be shown with title url and favicon.
- Searching using search bar will filter out the synced session information.
- Clicking item inside a synced session will be opening that link in a different tab.

As last part of synced session testing

- Open tabs in normal mode should be shown on other devices  (desktop browser lists this information under History -> Tabs from other devices).
- Changing the website on a tab etc should be visually observed on other devices with every sync cycle change update.

### Test Plan for Send - Receive Webpage:

Basically there are 2 things to test here 

Sending a webpage to another device

- Load url and press .. menu and find Send To Your Devices option available ( this option should be also available as an activity inside share sheet when Share with... is pressed.)
- Click Send To Your Devices option and check device which are capable of receiving a webpage is presented in that list with the device name, last active date and icon that shows what type of device it is.
- Cancel button should dismiss the screen and Send To Your Device should send the webpage information to selected device.
- Clicking different devices should choose which device to send the webpage and the checkmark update according to which device is selected.

Receiving a website from another device

Send a webpage from another device iOS/Android/Desktop.
Observe a Toast appears at the bottom after sync cycle is done written Tab Received and with Go action button.
Press Go button to open the link in a new tab.

### Some additional videos related with Test Plans mentioned here.

Open Tabs Tab Tray

https://user-images.githubusercontent.com/6643505/184030099-9aaa2f20-cb51-43a4-bffb-f87a3ae7fb3b.mov

Send Webpage

https://user-images.githubusercontent.com/6643505/184030149-cb89e98b-29b1-4555-b129-94edb1eb6f6c.MP4


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
